metrics.py

```diff
@@ -405,29 +436,77 @@ def __init__(self, exogenous_attr: Collection, indices_to_ignore=None):
         self._is_directed = True
 
     def _calc_edge_weights(self):
+        """
+        Each edge is weighted by the index of it's corresponding sorted type pair.
+        
+        For example, for an n=4 matrix with types [A, B, A, B], the weight matrix is
+            [ AA, AB, AA, AB
+              BA, BB, BA, BB
+              AA, AB, AA, AB
+              BA, BB, BA, BB ]
+        but instead of AA, AB, ..., we save the corresponding index of the type pair
+        in self._sorted_type_pairs (which is sorted alphabetically.)
+            
+        """
         num_nodes = len(self.exogenous_attr)
-        unique_types = sorted(set(self.exogenous_attr))
-        self.edge_weights = np.zeros((self._get_num_weight_mats(), num_nodes, num_nodes))
-        weight_mat_idx = 0
-        num_skipped_indices = 0
-        for pre_type in unique_types:
-            for post_type in unique_types:
-                if weight_mat_idx not in self._indices_to_ignore:
-                    for i in range(num_nodes):
-                        for j in range(num_nodes):
-                            if i == j:
-                                continue
-                            if self.exogenous_attr[i] == pre_type and self.exogenous_attr[j] == post_type:
-                                self.edge_weights[weight_mat_idx - num_skipped_indices, i, j] = 1
-                else:
-                    num_skipped_indices += 1
-                weight_mat_idx += 1
+        self.edge_weights = np.zeros((num_nodes, num_nodes)).astype(int)
+
+        for i in range(num_nodes):
+            for j in range(num_nodes):
+                type_1 = self.exogenous_attr[i]
+                type_2 = self.exogenous_attr[j]
+
+                self.edge_weights[i, j] = self._sorted_type_pairs.index((type_1, type_2))
 
     def _get_num_weight_mats(self):
         return len(set(self.exogenous_attr)) ** 2 - len(self._indices_to_ignore)
 
+    def calculate(self, input: np.ndarray):
+        nonzero_entries = input.nonzero()
+        type_pairs_in_network = self.edge_weights[nonzero_entries].astype(int)
+        return np.delete(np.bincount(type_pairs_in_network, minlength=len(self._sorted_type_pairs)), self._indices_to_ignore)
+        
     def calculate_for_sample(self, networks_sample: np.ndarray):
```

Comment:
Essentially, the problem here is to apply the bincount/histogram along a single axis (I don't get why numpy don't provide that). Because if we have this functionality, we can multiply each sample by the weight matrix (e.g. `np.einsum('ijk,ij->ijk', networks_sample, self._edge_weights)`, probably there is a better solution), and perform the counting along the third axis (a histogram per nXn matrix along the first 2 axes).
There is a dude here that says he knows to do that efficiently, but I don't really understand what he does... Let's think about that together.
https://stackoverflow.com/questions/44152436/calculate-histograms-along-axis